src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,370 @@
+   * produce-interpolants which is set here. From CVC5AbstractProver Line 66
+   */
+  private void setSolverOptions(int randomSeed, Set<ProverOptions> pOptions, Solver solvertoSet) {
+    if (incremental) {
+      solvertoSet.setOption("incremental", "true");
+    }
+    if (pOptions.contains(ProverOptions.GENERATE_MODELS)) {
+      solvertoSet.setOption("produce-models", "true");
+    }
+    if (pOptions.contains(ProverOptions.GENERATE_UNSAT_CORE)) {
+      solvertoSet.setOption("produce-unsat-cores", "true");
+    }
+    solvertoSet.setOption("produce-assertions", "true");
+    solvertoSet.setOption("dump-models", "true");
+    solvertoSet.setOption("output-language", "smt2");
+    solvertoSet.setOption("seed", String.valueOf(randomSeed));
+    solvertoSet.setOption("produce-interpolants", "true");
+
+    // Set Strings option to enable all String features (such as lessOrEquals)
+    solvertoSet.setOption("strings-exp", "true");
+
+    // Enable more complete quantifier solving (for more info see CVC5QuantifiedFormulaManager)
+    solvertoSet.setOption("full-saturate-quant", "true");
+  }
+
+  @Override
+  public Term addConstraint(BooleanFormula pConstraint) throws InterruptedException {
+    Preconditions.checkState(!closed);
+    Term t = creator.extractInfo(pConstraint);
+
+    super.addConstraint(pConstraint);
+    return t; // t is not returned in the Abstract Class
+  }
+
+  @Override
+  public BooleanFormula getInterpolant(Collection<Term> pFormulasOfA)
+      throws SolverException, InterruptedException {
+    Preconditions.checkState(!closed);
+
+    if (pFormulasOfA.isEmpty()) { // Catch trivial case
+      return mgr.getBooleanFormulaManager().makeBoolean(true);
+    }
+
+    // formulasOfB := assertedFormulas - pFormulasOfA
+    Set<Term> formulasOfB =
+        assertedFormulas.stream()
+            .flatMap(c -> c.stream())
+            .filter(n -> !pFormulasOfA.contains(n))
+            .collect(ImmutableSet.toImmutableSet());
+
+    if (formulasOfB.isEmpty()) { // Catch trivial case
+      return mgr.getBooleanFormulaManager().makeBoolean(false);
+    }
+
+    // fit the Input to work with getCVC5Interpolation
+
+    ImmutableList<Collection<Term>> formAAsList = ImmutableList.of(pFormulasOfA);
+    ImmutableList<Collection<Term>> formBAsList = ImmutableList.of(formulasOfB);
+
+    Term itp = getCVC5Interpolation(formAAsList, formBAsList);
+
+    return creator.encapsulateBoolean(itp);
+  }
+
+  @Override
+  public List<BooleanFormula> getSeqInterpolants(
+      List<? extends Collection<Term>> partitionedFormulas)
+      throws SolverException, InterruptedException {
+    Preconditions.checkArgument(
+        !partitionedFormulas.isEmpty(), "at least one partition should be available.");
+
+    List<BooleanFormula> itps = Stream.of(new BooleanFormula[] {}).collect(Collectors.toList());
+    for (int i = 1; i < partitionedFormulas.size(); i++) {
+      itps.add(
+          getInterpolant(
```

Comment:
I am still missing a proof that this interpolation technique delivers sequential interpolants. 
The interpolation strategy does not enforce a consistent internal state (e.g., internal proof) and seems to even build up the solver-stack several times.

If you are already pushing and popping on a solver-stack, please reuse the first sequential interpolants for further computation. This makes the later computation potentially faster, because the existing interpolants reduce workload.

Reply:
I took the implementation of `Mathsat5InterpolationProver`. The only comment:
```
Mathsat5InterpolatingProver line 136: the fallback to a loop is sound and returns an inductive sequence of interpolants
```
it sounds sketch but i went with it. 
Could you elaborate on how the interpolation pairs would look, if the already calculated sequential interpolants would be used for further interpolations? 
Lets say we have following partitionedFormulas `[A,B,C,D,E,F]`
1. `itp0 = getInterpolation([A],[B,C,D,E,F])`
2. `itp1 = getInterpolation([itp0],[C,D,E,F])`
3. ...
Something like this?

Reply:
You did correctly got the idea for sequential interpolation.

For an SMT solver that uses the same internal proof for all interpolants, it is easy to provide valid sequential interpolants. (It actually is not "easy". We reported some bugs to the solver-developers in the early times, until they got it right... or removed interpolation support from the solver :smile:). 

I found some colorful slides in an archived folder (I am happy to use them again, after 8 years :smile: no guarantee for correctness! ):

The colors denote the first (GREEN) and second (RED) part of the interpolation query, and the interpolant (BLUE). 

![image](https://github.com/sosy-lab/java-smt/assets/10998542/86f29cd7-ad04-480e-b35c-bb302fb47289)

![image](https://github.com/sosy-lab/java-smt/assets/10998542/1939cd49-4c58-4dea-a955-e0a25e805c93)
![image](https://github.com/sosy-lab/java-smt/assets/10998542/6340afc5-5cb6-4408-8179-813728b32e25)
![image](https://github.com/sosy-lab/java-smt/assets/10998542/6e015089-740b-4a36-9bac-ab18a43c3a33)

![image](https://github.com/sosy-lab/java-smt/assets/10998542/94b2445d-607e-48f6-89c6-618105f6ee44)
![image](https://github.com/sosy-lab/java-smt/assets/10998542/1d267981-9208-4e91-bce9-bbbe94b32d23)
![image](https://github.com/sosy-lab/java-smt/assets/10998542/a8a3efd2-e279-4f59-9b58-e3b5bf759fe3)
![image](https://github.com/sosy-lab/java-smt/assets/10998542/01eeeb1e-4057-4bbe-84a2-01be5dd841b7)
![image](https://github.com/sosy-lab/java-smt/assets/10998542/c6c900c9-ccb3-4345-af22-efbe7a2c973e)

Reply:
I implemented it the way you suggested. Surprisingly :) the tests do not fail anymore. Thanks a lot! With this approach, no further prove is needed right?